Don't throw exceptions on the replica side when handling single partition reads and writes

piodul opened the pull request:
This PR gets rid of exception throws/rethrows on the replica side for writes and single-partition reads. This goal is achieved without using `boost::outcome` but rather by replacing the parts of the code which throw with appropriate seastar idioms and by introducing two helper functions:

1.`try_catch` allows to inspect the type and value behind an `std::exception_ptr`. When libstdc++ is used, this function does not need to throw the exception and avoids the very costly unwind process. This based on the "How to catch an exception_ptr without even try-ing" proposal mentioned in https://github.com/scylladb/scylla/issues/10260.

This function allows to replace the current `try..catch` chains which inspect the exception type and account it in the metrics.

Example:

```c++
// Before
try {
    std::rethrow_exception(eptr);
} catch (std::runtime_exception& ex) {
    // 1
} catch (...) {
    // 2
}

// After
if (auto* ex = try_catch<std::runtime_exception>(eptr)) {
    // 1
} else {
    // 2
}
```

2. `make_nested_exception_ptr` which is meant to be a replacement for `std::throw_with_nested`. Unlike the original function, it does not require an exception being currently thrown and does not throw itself - instead, it takes the nested exception as an `std::exception_ptr` and produces another `std::exception_ptr` itself.

Apart from the above, seastar idioms such as `make_exception_future`, `co_await as_future`, `co_return coroutine::exception()` are used to propagate exceptions without throwing. This brings the number of exception throws to zero for single partition reads and writes (tested with scylla-bench, --mode=read and --mode=write).

Results from `perf_simple_query`:

```
Before (719724e4df6b12524a6428f0a9ab20027bbed254):
  Writes:
    Normal:
      127841.40 tps ( 56.2 allocs/op,  13.2 tasks/op,   50042 insns/op,        0 errors)
    Timeouts:
      94770.81 tps ( 53.1 allocs/op,   5.1 tasks/op,   78678 insns/op,  1000000 errors)
  Reads:
    Normal:
      138902.31 tps ( 65.1 allocs/op,  12.1 tasks/op,   43106 insns/op,        0 errors)
    Timeouts:
      62447.01 tps ( 49.7 allocs/op,  12.1 tasks/op,  135984 insns/op,   936846 errors)

After (d8ac4c02bfb7786dc9ed30d2db3b99df09bf448f):
  Writes:
    Normal:
      127359.12 tps ( 56.2 allocs/op,  13.2 tasks/op,   49782 insns/op,        0 errors)
    Timeouts:
      163068.38 tps ( 52.1 allocs/op,   5.1 tasks/op,   40615 insns/op,  1000000 errors)
  Reads:
    Normal:
      151221.15 tps ( 65.1 allocs/op,  12.1 tasks/op,   43028 insns/op,        0 errors)
    Timeouts:
      192094.11 tps ( 41.2 allocs/op,  12.1 tasks/op,   33403 insns/op,   960604 errors)
```

piodul commented:
cc: @avikivity @elcallio @nyh @psarna 

I would like to hear you opinions about this. This approach seems much nicer to me than performing "resultification" and requires less effort. Unfortunately, I don't know how to make it work with libc++ - internals responsible for performing something similar to `__do_catch` are well hidden - but I'm not sure if it is a problem as we are currently using libstdc++.

We can also implement `handle` similar to the one in the proposal, but for now I'm just sending `try_catch` as it is sufficient for demonstration purposes.

piodul commented:
cc: @gleb-cloudius

avikivity commented on utils/abi/libstdc++/unwind-cxx.h:
```diff
@@ -0,0 +1,401 @@
+//
+// This file is part of GCC.
+//
+// GCC is free software; you can redistribute it and/or modify
```
We can't incorporate this due to license conflict (the licenses are compatible, but the original contributor didn't sign our copyright license agreement). Maybe you can pick off the original noted below.

piodul replied:
I wrote a new header from scratch with necessary utilities after reading the Itanium ABI documentation.

avikivity commented on utils/abi/libstdc++/unwind-cxx.h:
```diff
@@ -0,0 +1,401 @@
+// <http://www.gnu.org/licenses/>.
+
+// This is derived from the C++ ABI for IA-64.  Where we diverge
+// for cross-architecture compatibility are noted with "@@@".
```
Here, if the license is more permissive and if it didn't diverge in an important way.

avikivity commented on utils/handle_exception.hh:
```diff
@@ -0,0 +1,58 @@
+        return reinterpret_cast<T*>(opt_ptr);
+    }
+    return nullptr;
+#else
```
The fallback makes it more palatable.

avikivity commented:
"interesting" (which means, I don't know what my opinion is).

In terms of code churn it's a mixed blessing: one one hand less churn and less result<> intrusiveness is welcome. On the other hand it is now easy to introduce a code path that isn't fast-exception-aware and so introduce a regression. I suppose a performance regression test that triggers timeouts intentionally (by using a fake slow disk and BYPASS CACHE or similar) is not  hard to add.

avikivity commented on utils/handle_exception.hh:
```diff
@@ -0,0 +1,58 @@
+    } catch (T& t) {
+        return &t;
+    } catch (...) {
+        return nullptr;
```
It's moot even without the optimizer (and this is fallback code, _and_ it's in a try-catch that is inherently super slow)

piodul replied:
I'm not sure what the problem is here. Both `catch` blocks are necessary. In the fallback implementation, the `try_catch` function throws the exception to check whether it matches the type `T`. The thrown exception is not supposed to exit the `try_catch` function, so we need the `catch (...)` part.

I moved the `return nullptr` after the `catch(...) {}` part, but it probably doesn't make any difference

denesb commented on service/storage_proxy.cc:
```diff
@@ -2785,7 +2785,12 @@ void storage_proxy::send_to_live_endpoints(storage_proxy::response_id_type respo
             }
 
-            if (coordinator == my_address) {
+            if (handler.get_schema()->cf_name() == "tbl_noop") {
```
We can use the error injection framework for injection errors.

piodul replied:
Yes, those are hacks I included only for demonstration purposes of the RFC. I agree we shouldn't leave them in the final version. I sent a separate PR which adds flags to `perf_simple_query` so that you can trigger timeouts and measure their performance without such hacks.

The PR is here: https://github.com/scylladb/scylla/pull/10899

piodul commented:
v0 -> v1:

- Changed the title and reworded the description. The PR is still based on the concept of the non-throwing "try_catch", its scope was extended a bit but I hope it is more precise.
- Replaced `unwind-cxx.h` with a header implemented based on the Itanium ABI description.
- Moved `try_catch` out of the `utils` namespace, and into the `utils/exceptions.hh` file.
- Added `make_nested_exception_ptr` function.
- Added tests for `try_catch` and `make_nested_exception_ptr`, for both the optimized and fallback implementations.
- Removed the "hacked in" tables `tbl_noop` i `tbl_fail` tables.
- Got rid of other throws and rethrows on the replica path for writes and single partition reads.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1162/

psarna commented on utils/exceptions.cc:
```diff
@@ -73,3 +74,25 @@ bool is_timeout_exception(std::exception_ptr e) {
+#if defined(__GLIBCXX__)
+
+#include <typeinfo>
+#include "utils/abi/eh_ia64.hh"
```
Looks like we should also add another #ifdef for checking target architecture and fall back to previous code if it isn't compatible. Scylla also compiles on aarch64, which might follow a different ABI? In which case we should neither compile these snippets nor enable throwless catch in the first place.

avikivity replied:
We should probably opt-in to the optimization by arch here.

But:
 - please check aaaaaaaaaaaaaarch64 now and opt-in if it works
 - emit a #warn for archs that didn't opt-in, since they will suffer significant degradation (we have -Wno-error=cpp)

avikivity commented on utils/exceptions.hh:
```diff
@@ -60,3 +67,36 @@ inline void maybe_rethrow_exception(std::exception_ptr ex) {
+
+} // utils::internal
+
+/// If the exception_ptr holds an exception which would a `catch (T&)` clause,
```
A word is missing after "which would"

piodul commented:
> We can also consider donating it to Seastar, if we see fast-path exceptions there.

I scanned the seastar code and I couldn't find try..catches on the fast path. The only place I find slightly suspicious is [here](https://github.com/scylladb/seastar/blob/024ac1785511917c139b2410db35e4e525618edd/include/seastar/rpc/rpc_impl.hh#L607) - IIUC if rpc tries to send data faster than it is possible, a large volume of timeouts might be generated and each one of them will be rethrown once in that `handle_exception_type`.

There is `future::handle_exception_type` which would benefit from `try_catch` (although in Scylla we can just convert them to `future::handle_exception` + `try_catch` with the same effect).

We also have `operator<<` for `std::exception_ptr`, which rethrows the exception three times. Those throws could be removed with `try_catch`.

avikivity commented:
> > We can also consider donating it to Seastar, if we see fast-path exceptions there.
> 
> I scanned the seastar code and I couldn't find try..catches on the fast path. The only place I find slightly suspicious is [here](https://github.com/scylladb/seastar/blob/024ac1785511917c139b2410db35e4e525618edd/include/seastar/rpc/rpc_impl.hh#L607) - IIUC if rpc tries to send data faster than it is possible, a large volume of timeouts might be generated and each one of them will be rethrown once in that `handle_exception_type`.


A common case when this happens is when a node dies. Then you get an exception storm on the clients. So this is the main target. But let's gain experience with it first.

> 
> There is `future::handle_exception_type` which would benefit from `try_catch` (although in Scylla we can just convert them to `future::handle_exception` + `try_catch` with the same effect).

Yes.

> 
> We also have `operator<<` for `std::exception_ptr`, which rethrows the exception three times. Those throws could be removed with `try_catch`.

Heavy duty exceptions like timeouts shouldn't be printed. So this one is unimportant.

We remove our timeout printouts btw. Log storms are also slow.

piodul commented:
> @piodul did these exception inspection helpers ever get fast libc++ versions?

@travisdowns no, we have been sticking to libstdc++ so there haven't been any incentive for us to implement them.
